// gl-react-native/lib/ImageSourceTextureLoader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _glReact = require("gl-react");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function imageSourceHash(imageSource) {
  if (typeof imageSource === "number") return imageSource;
  var uri = imageSource.uri;

  if (!uri) {
    throw new Error("GLImages: unsupported imageSource: {uri} needs to be defined");
  }
  return uri;
}

var ImageSourceTextureLoader = function (_TextureLoader) {
  _inherits(ImageSourceTextureLoader, _TextureLoader);

  function ImageSourceTextureLoader() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ImageSourceTextureLoader);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ImageSourceTextureLoader.__proto__ || Object.getPrototypeOf(ImageSourceTextureLoader)).call.apply(_ref, [this].concat(args))), _this), _this.loads = new Map(), _this.textureAssets = new Map(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ImageSourceTextureLoader, [{
    key: "dispose",
    value: function dispose() {
      this.loads.forEach(function (d) {
        d.dispose();
      });
      this.loads.clear();
      var gl = this.gl;

      var rngl = gl.getExtension("RN");
      this.textureAssets.forEach(function (_ref2) {
        var texture = _ref2.texture;

        rngl.unloadTexture(texture);
      });
      this.textureAssets.clear();
    }
  }, {
    key: "canLoad",
    value: function canLoad(input) {
      return typeof input === "number" || input && (typeof input === "undefined" ? "undefined" : _typeof(input)) === "object" && typeof input.uri === "string";
    }
  }, {
    key: "load",
    value: function load(imageSource) {
      var _this2 = this;

      var hash = imageSourceHash(imageSource);
      var load = this.loads.get(hash);
      if (load) return load;
      var ignored = false;
      var dispose = function dispose() {
        ignored = true;
      };
      var promise = this.gl.getExtension("RN").loadTexture({ yflip: true, image: imageSource }).then(function (data) {
        _this2.textureAssets.set(hash, data);
        return data.texture;
      });
      var d = { dispose: dispose, promise: promise };
      this.loads.set(hash, d);
      return d;
    }
  }, {
    key: "get",
    value: function get(imageSource) {
      var asset = this.textureAssets.get(imageSourceHash(imageSource));
      return asset && asset.texture;
    }
  }, {
    key: "getSize",
    value: function getSize(imageSource) {
      var asset = this.textureAssets.get(imageSourceHash(imageSource));
      if (!asset) return;
      return [asset.width, asset.height];
    }
  }]);

  return ImageSourceTextureLoader;
}(_glReact.TextureLoader);

exports.default = ImageSourceTextureLoader;
